### ISSUE TITLE
API Rewrite?

### ISSUE THREAD

**Issue by Owez** [braincore/pid-rs#8]
I'm currently learning about control systems and wanted to implement a PID system as a starting project and found it was already done here. I used this project as a way to start to reimplement a new version of `pid-rs` privately to learn and figured that the API might as well be improved upstream to make it worthwhile :)

How interested would this project be with an API rewrite, looking something like this currently:

```rust
use pid::Controller;

let mut controller: Controller<f64> = Controller::basic(100.0).p(10.0, 2.5).i(666.0, None);
println!("Current controller: {:?}", controller);

loop {
    let output = controller.calc(2.75);
    println!("Output: {:?}", output);
}
```

**Comment by braincore** [comment]
Thanks for reaching out!

I'd characterize this as a builder-like pattern. I'm primarily a fan of it when there are a large number of arguments and only a small subset of them are expected to be used. In the case of `Pid`, it's likely that all parameters need to be specified so this serves more as a pure-aesthetic-upgrade that compensates for rust's lack of named parameters.

Overall, while this library isn't popular, it does appear to be used by others and I'd prefer not to break their builds for something purely aesthetic (only if the library got very crufty).

Then again, I haven't been following the latest rust conventions. Has this pattern become more popular?

**Comment by Owez** [comment]
Yeah that's fair, the changes I've thought about so far where:

```rust
/// Function-based
use peaid::pid;

fn main() {
    let controller_pi = pi(0.1, None, 10.0, 5.0);
    let controller_pid = pid(0.0, None, 5.5, 2.3, 10.0, None);
}

/// Current thinking
use peaid::Controller;

fn main() {
    let controller_pd = Controller::new().p(0.0, None).d(0.0);
    let controller_pid = Controller::new().p(0.0, None).i(5.5, 2.3).d(10.0, None);
}

/// Macros (not good for production)
#[macro_use]
use peaid::pid;

fn main() {
    let controller_id = pid!(f64; i: 0.1, d: {5.5, 2.3});
    let controller_pid = pid!(f64; p: 2.5, i: {2.5, 0.5}, d: {5.5, 2.3});
}

/// Hybrid
use peaid::Controller;

fn main() {
    let controller_id = Controller::new_id(0.5, None, 2.5, 5.75);
    let controller_pid = Controller::new_pid(10.5, None, 10.0, None, 11.25, 75.2);
}
```

And a couple of friends agreed on the setter pattern, though I'm not sure how popular it is. The main library I can remember using it was clap before it went with the structopt based API. The main reason I originally went for the setter pattern because it was the simplest way I could think of as the rewrite allows any combo of p, i and d -- that and some other small naming changes would probably break backwards compat unfortunately.

**Comment by quietlychris** [comment]
Personally, I'd be interested in a version of this library with a Builder-type instantiation pattern--I've considered using this library before, but I honestly feel a little overwhelmed building a struct with 8 un-named fields. I know theoretically the argument pattern makes sense, but for some reason it doesn't work well in my brain. Maybe something like 

```rust
let mut controller: Pid = Pid::default(p_term, i_term, d_term).build(); // Make the limits terms auto-calculated based on supplied term
// or maybe something like
let mut controller: Pid = Pid_Builder::default(p, i, d).set_limts(p_limit, i_limit, d_limit).set_setpoint(value).build();

// Which could be useful for features in the future
let mut controller: Pid = Pid_Builder::default(p, i, d).store_history(true).build();
controller.auto_tune(); // Requires store_history() to have been turned on

// Manual access to all public fields, sorta similar to the existing reset_integral_term() function
controller.limits(p_limit, i_limit, t_limit);
controller.p_limit(p_lim_term);
controller.setpoint(new_setpoint);
```
I'm currently using my own basic implementation of a PID controller, but am definitely interested in switching over to something more robust at some point in the future.

**Comment by braincore** [comment]
Okay, let's change it.

Requirements that I'd like to maintain:

- Use `next_control_output()` instead of `calc()` because I value using names from PID literature rather than generic names.
- Limits shouldn't be optional. PIDs are often used with actuators and an engineer needs to consider them. They can make them large at their own peril.
- Minimize renaming.

@quietlychris `build()` is favorable once there are immutable configurations like `store_history()`. But, as of now, every field is mutable, so we might as well maintain symmetry between configuration during creation and configuration while running. I'd rather not future-proof this.

So, going off of @Owez's initial proposal:

```
use pid::Pid;

// on creation, kp = kd = ki = setpoint = 0
// also, k_limit = d_limit = i_limit = 0
let mut pid = Pid::new(OUTPUT_LIMIT)
  .p(KP, P_LIMIT)
  .d(KD, D_LIMIT)
  .i(KI, I_LIMIT)
  .setpoint(SETPOINT);

// at any later point
pid.p(KP, P_LIMIT)
pid.d(KD, D_LIMIT)
pid.i(KI, I_LIMIT)
pid.setpoint(SETPOINT)
```

Any further thoughts? @quietlychris: Would this have been a sufficiently friendly/appealing interface to you? Also, I saw that you have a quadcopter repo. I built this PID for my quadcopter :). I have some suggestions about your HW choices that I'll post in your repo.

@Owez, if you want to submit a PR, happy to review.

**Comment by Owez** [comment]
Looks good so far :)

Perhaps `setpoint` could be integrated into the `Pid::new()` or the output limit be integrated into the `.setpoint()` pattern for consistency? I see how it makes sense with the config you posted though.

I'll have the PR done in a couple days; the next step after this would be to add some sort of auto-calculator?

**Comment by braincore** [comment]
Re: auto-tuning: I suspect the interface for this won't be part of Pid, but instead the Pid will be composed in an `AutoPid` which will manage the pid's parameters. Do you have any strategies for auto-tuning in mind? It's a difficult problem, and personally I resorted to manual tuning for a quadcopter.

An auto-tuner would have been nice in my application because as the battery drained (causing voltage drop across motors), the optimal pid characteristics changed.

**Comment by quietlychris** [comment]
@braincore I certainly understand your point about future-proofing it. And yes, at least on my end, I definitely find your suggested API change friendlier (at least to my eye, anyway). I also really like that it's `no_std` compatible--as you've seen, my quadcopter repo is all on Linux, so I have easy access to an allocator, but I think it's a good option. Your point about making limits required is also well-taken. 

I don't really have any ideas about auto-tuning at the moment; while it's a feature I'd definitely play with if it was available, I've also been relying on hand-tuning the gains for the moment. It's interesting to know that you saw a lift change as the battery voltage dropped--that's something that I'll have to keep an eye out for.

**Comment by Owez** [comment]
To be honest, I don't really know anything about autotuning but I have heard a couple people mention it as being a feature they'd like. I'll have a look at some resources on it, this is about as much as I understand haha:

![image](https://user-images.githubusercontent.com/30081702/118972998-19c55700-b969-11eb-88e8-a115c4e10a5d.png)

Perhaps before any autotuner is in place, having the `Measurements occur at equal spacing. (t(i) = t(i-1) + C)` assumption optionally fixed could be easier?

**Comment by quietlychris** [comment]
@braincore Just following up on this again, as I believe there's a PR (#9) by @Owez that addresses this! I'm definitely interested in using this library with the updated API soon, so while I don't want to be a pain (I'm sure you're super busy), I'd be really appreciative if you might be able to find some time over the next few days to look over the final merge :)

**Comment by ahepp** [comment]
Hi all,

Would be interested in seeing some kind of more ergonomic API move forward.

I'm very much a rust novice, but I put together a concept I had for a way to make this library more ergonomic, without breaking the existing API:

https://github.com/ahepp/pid-rs/commit/f671e28650eaf0e9515f349f89cef632ccd1368e

Apologies if this contains any footguns, I'm pretty new to rust.

**Comment by Owez** [comment]
> Hi all,
> 
> Would be interested in seeing some kind of more ergonomic API move forward.
> 
> I'm very much a rust novice, but I put together a concept I had for a way to make this library more ergonomic, without breaking the existing API:
> 
> ```
> let gains = PidGains::new(2.0, 0.0, 0.0);
> let limits = PidLimits::new(100.0, 100.0, 100.0, 100.0);
> let builder = PidBuilder::new(gains, limits, 10.0);
> let mut pid = Pid::from(builder);
> ```
> 
> A full patch with a test is available at [ahepp@f671e28](https://github.com/ahepp/pid-rs/commit/f671e28650eaf0e9515f349f89cef632ccd1368e)
> 
> Apologies if this contains any footguns, I'm pretty new to rust.
> 
> Of course one could also do this the other way:
> 
> ```
> let p = PidFactor::new(2.0, 100.0);
> let i = PidFactor::new(0.0, 0.0);
> let d = PidFactor::new(0.0, 0.0);
> let builder = PidBuilder::new(p, i, d, 10.0);
> let mut pid = Pid::from(builder);
> ```

The PID factor bit in the second codeblock is quite a neat idea, I'm thinking it could be implicitly made with ranges, so like `2.0..100.0` which could be cool. Waiting on the #9 PR to be looked at atm

**Comment by ahepp** [comment]
I noticed #9 and wasn't sure if it was on hold due to breaking the API. Thanks for contributing the patch. I think I prefer your API to either option I proposed.

@braincore, are you still interested in these API updates? If you're open, but busy with other projects, maybe we can look for additional maintainers? Appreciate the nice work maintaining the crate so far.

**Comment by braincore** [comment]
Hey everyone! Sorry for the delay in addressing this especially to @Owez for the rudeness of dropping off. I have one last minor request in the PR regarding the README.

Two follow up changes are needed:
1. Any updates to modernize the crate for the latest versions of rust. I haven't done much rust in the last two years.
2. Bump the version & publish.

Regarding @ahepp's on point suggestion to add maintainers, @Owez: Happy to make you a maintainer if you'd like, and ideally you could complete the two follow ups.

**Comment by Owez** [comment]
Suprisingly the two dependencies on the crate are still up to date :)

I bumped the version to v4 in be3b03a. I'd be happy to become a maintainer, been wanting to get into some more OSS work recently.

**Comment by braincore** [comment]
@Owez Awesome. You'll need to create an account on [crates.io](https://crates.io/) so that I can add you as an owner to publish a new version.

**Comment by Owez** [comment]
> @Owez Awesome. You'll need to create an account on [crates.io](https://crates.io/) so that I can add you as an owner to publish a new version.

I've got my account on crates.io at https://crates.io/users/Owez, is there any sort of link on crates.io/github that you need?

**Comment by braincore** [comment]
@Owez Sent an invite

**Comment by Owez** [comment]
> @Owez Sent an invite

Thanks, accepted

**Comment by ua-kxie** [comment]
Good to see renewed life here. 

While I thought this was dead I went off and wrote my own [controller lib](https://github.com/ua-kxie/pid-ctrl) using this one as a reference.

As for autotuning, there seems to be an [accepted procedure](https://www.sciencedirect.com/topics/computer-science/ziegler-nichols-method) which is very straightforward. The main problem is detecting instability/oscillation and then determining its frequency. The process doesn't involve the pid controller, so the autotuner can (and probably should) be its own project.

**Comment by braincore** [comment]
@Owez I'll close this issue once you've published v4.

**Comment by braincore** [comment]
@ua-kxie Nice. Perhaps with @Owez being more responsive, you can fold some of those changes back into this crate.

For my application (quadcopter), the ZN Method was definitely a good starting point, but there were two things I noticed:

1. Manual tweaking could produce noticeably better I and D terms. I had a live stream of 100+Hz (sampled) gyro + accelerometer that I'd used for calibration.
2. The level of non-linearity and instability in quadcopters made it such that different PID parameters should be used in different situations where a situation would be defined by factors such as attitude (yaw/pitch/roll), the motor power (related to battery level as li-ion voltage drops as it discharges), and proximity to the ground ([ground effect](https://en.wikipedia.org/wiki/Ground_effect_(aerodynamics))).

Thinking out loud, #1 is probably better addressed in a separate crate. #2 could be in a separate crate that utilizes this one.

**Comment by braincore** [comment]
@ahepp Regarding your proposal, I like that you were considerate of maintaining backwards compatibility. In general, however, I have less appetite for aesthetic improvements that add more abstraction without a burning reason to.

I liked @Owez's approach because it matched my experience that configuring and updating a PID often happens on a per-term (P, I, D) basis, and so it makes sense to have the API share that mental model even if in the end the coding isn't much different.

**Comment by Owez** [comment]
Added some final docs and published. I think ideally we could document out the issue of autocalibration by pointing to a crate with it inside. If there was an abstract autocalibrator we could always implement a pid-rs binding and feature-gate it but I've not looked into what's available.

**Comment by braincore** [comment]
@Owez It's good to push a git tag and mark a new release in GH with release note (best to copy and paste what you've already written). As annoyingly redundant as it is, I appreciate it as a consumer of other libraries.

**Comment by Owez** [comment]
> @Owez It's good to push a git tag and mark a new release in GH with release note (best to copy and paste what you've already written). As annoyingly redundant as it is, I appreciate it as a consumer of other libraries.

Will do, noticed the releases yesterday but forgot about it today
